### PR TITLE
Contact lists

### LINKED ISSUE #5407: Manage contact lists
## User story
As a user I want to be able to create recipient lists so that i can assign multiple email address to a single identifier.

## AC

* There is a new section in the folder of the contact view named "Recipient lists"
* Next to heading there is an "Add" button.
* A dialog appears where the user can specify a name for the recipient list.
* The dialog also contains an address input field to add recipients that behaves similar to adding guests to an event.
* An added recipient is shown below the address input  field.
* second column shows a list of all recipients in the selected list and supports multiselect
* if only one recipient is selected, the third column shows a view with
* card that shows the recipient address and a "remove from list" button and a "create contact" button that brings up a contact * editor with that address prepopulated
* one card per contact in the current users contact list that has the recipient address assigned as a full contact viewer. each has delete and edit in the more menu
* in multiselect, a multiselection viewer is shown similar to the mail multiselect:
   * cancel seelction
   * mass actions (remove from list)
   * delete recipient lists is possible similar to calendars
* edit recipient list: renaming, adding items is possible
* Recipient lists are only available for new paid plans 

## Notes
- Service similar to calendar or templates (keep in mind sharing in the future)

## Tasks
- [x] Make a model for recipient list that has entries of email
- [x] Make a service that can create, delete, and edit recipient lists
- [x] Add section for recipient list in contact view
- [x] Create list view (2nd column)
- [x] Be able to add new recipient lists (create dialog)
- [x] Be able to remove recipient lists
- [x] Create recipient list viewer (3rd column view)
- [x] Be able to add new entries to the list (create dialog)
- [x] Remove entries from list
- [x] Have create contact button on emails that do not have a contact
- [x] Add support for multiselection
- [x] Edit recipient list
- [x] Make recipient list are only available for paid plans (check in to see which plan)

## Test notes
TBD

 - [ ] Test migration from previous model for customer with and without custom plan (to get custom plan upgrade an account from old plan and then roll it back in the admin client)
 - [ ] Creating contact lists is possible with new paid plans
 - [ ] Creating contact lists is not possible with free plan
 - [ ] Creating contact lists is not possible with old plans
 - [ ] Downgrading is not possible if there are any contact lists
 - [ ] Opening contact view loads contact lists
 - [ ] Deleting contact list is possible
 - [ ] Creating contact list entries is possible
 - [ ] Renaming contact list is possible
 - [ ] Showing contact lists is possible
   - [ ] Desktop
   - [ ] Mobile
 - [ ] Related contacts are shown in contact list entry viewer
     - [ ] you can edit contact in in this view
     - [ ] you can delete contact in this view
 - [ ] If there is no related contact, "No contacts with this email found. Create Contact" message shows
     - [ ] clicking "create contact" brings up the contact editor with the email address already filled in
     - [ ] you can save and contact appears 
 - [ ] Creating/deleting entry updates the list
 - [ ] Loading entry by URL will load and select the entry
 - [ ] Switching between lists/contacts works correctly and updates the URL
 - [ ] Multiselect actions (delete) work
 - [ ] clicking the plus button in the list toolbar brings up dialog where more recipients can be added
     - [ ] a recipient already in the list (in the dialog) will not be added again
     - [ ] a recipient already in the greater contact list will be added in the dialog, but after saving there will only be one entry
 - [ ] Actions in Mobile view are appropriate to Contact or Contact List
     - [ ] Clicking the Plus button in the header will add Contact or Contact List Entries 
     - [ ] The sort option in the header shows and works for contacts, but does not display when looking at a contact list   

Regression
- [ ] Interacting with contacts through "All Contacts" still works
    - [ ] Multiselect actions work
- [ ] You can use "New Contact" button to create a new contact
- [ ] loading contact by URL will load and select contact
- [ ] Contact actions (delete, edit, export) still work as expected

### PR DESCRIPTION
**Pull request by wrdhub**
close #5407

### CODE REVIEW

**Review by ganthern — COMMENTED** [review]
seems a bit as if contactView and ContactListView could be separate classes, there are a *lot* of inContactListView calls

File: src/contacts/ContactListEditor.ts [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+
+	addRecipient(address: string) {
+		if (!this.addresses.includes(address)) {
+			this.addresses = [address, ...this.addresses]
+		}
+	}
+
+	removeRecipient(address: string) {
+		this.addresses = this.addresses.filter((a) => address !== a)
+		m.redraw()
```
[the remark this answers is not in the document]
Comment (ganthern) [reply src/contacts/ContactListEditor.ts:108]: still think this should be changed, there's not much to test atm on this model, but once we do the mithril import will cause problems.

might also want to add a unit test file to make sure the deduplication works and to make it easier on the first person that needs to change this thing

File: src/translations/en.ts [hunk 2]
```diff
@@ -1538,6 +1538,15 @@ export default {
 		"yourCalendars_label": "Your calendars",
 		"yourFolders_action": "YOUR FOLDERS",
 		"yourMessage_label": "Your message",
-		"you_label": "You"
+		"you_label": "You",
+		// todo: add to phrase
```
Comment (ganthern) [inline src/translations/en.ts:1542]: todo

File: src/misc/TranslationKey.ts [hunk 3]
```diff
@@ -1524,3 +1524,13 @@ export type TranslationKeyType =
 	| "yourMessage_label"
 	| "you_label"
 	| "emptyString_msg"
+	| "cannotAddAttendees_msg"
+	// todo: add to phrase
```
Comment (ganthern) [inline src/misc/TranslationKey.ts:1528]: todo

File: src/contacts/view/ContactViewer.ts [hunk 4]
```diff
@@ -67,6 +70,38 @@ export class ContactViewer implements ClassComponent<ContactViewerAttrs> {
 						),
 						this.hasBirthday(contact) ? m("", this.formattedBirthday(contact)) : null,
 					]),
+					contact && attrs.editAction && attrs.deleteAction
```
Comment (ganthern) [inline src/contacts/view/ContactViewer.ts:73]: i assume these are guaranteed to always both be null or both be present?

File: src/contacts/view/ContactViewer.ts [hunk 5]
```diff
@@ -67,6 +70,38 @@ export class ContactViewer implements ClassComponent<ContactViewerAttrs> {
...
+											title: "more_label",
+											icon: Icons.More,
+										},
+										childAttrs: () => {
+											return [
+												{
+													label: "edit_action",
+													icon: Icons.Edit,
+													click: () => {
+														assertNotNull(attrs.editAction)(contact)
```
Comment (ganthern) [inline src/contacts/view/ContactViewer.ts:89]: adding a static message to the assertNotNull call makes it even easier to find if something goes wrong

File: src/contacts/view/ContactListViewModel.ts [hunk 6]
```diff
@@ -0,0 +1,254 @@
...
+		}
+	}
+
+	updateSelectedContactList(selected: Id): void {
+		this.selectedContactList = selected
+		this.listModel?.loadInitial()
+	}
+
+	updateContactList(contactListInfo: ContactListInfo, name: string, addresses: string[]): void {
+		// the name is stored on both GroupInfo (own calendar) and UserSettingsGroupRoot (shared with us calendar)
```
Comment (ganthern) [inline src/contacts/view/ContactListViewModel.ts:215]: this is probably not about calendars

File: src/api/worker/offline/OfflineStorageMigrator.ts [hunk 7]
```diff
@@ -24,6 +24,7 @@ import { sys86 } from "./migrations/sys-v86.js"
 import { sys87 } from "./migrations/sys-v87.js"
 import { sys88 } from "./migrations/sys-v88.js"
 import { tutanota62 } from "./migrations/tutanota-v62.js"
+import { sys89 } from "./migrations/sys-v89.js"
```
Comment (ganthern) [inline src/api/worker/offline/OfflineStorageMigrator.ts:27]: unused import?

**Review by ganthern — APPROVED** [review]

**Comment by wrdhub** [comment]
PR #5791 includes all these changes and should be the only one merged in.

**Pull request closed without merging**